### PR TITLE
Added possibility to add vertical lines to plot_raw.

### PR DESCRIPTION
**Pull request by jaeilepp**
Relating to discussion at #1206.
Added possibility to add vertical line to plot_raw and paint the area between. It is done by holding shift and clicking with left and right mouse buttons. Now we should decide if we want to do something with the painted area (possibility to crop etc.).
The logic also changed, so that now the bad channels and clicks to scrollbars can also be done with right mouse button. Wasn't sure if this is a bug or a feature so I let it be :)
There is also some redundancy with the old vertical line which should probably be removed...
raw_plot

### CODE REVIEW

**Comment by mainakjas** [comment]
Cool, great start! It seems that the old vertical line doesn't work and also the text associated with it. Maybe that function should still be retained? Basically, both should work. Should we mark the range associated with the area selection? Wdyt?

**Comment by mainakjas** [comment]
Also, we should have some way to unselect the area once it is selected, no?

**Comment by mainakjas** [comment]
Another commend, it would be good to show the selected area in the scrollbar panel. That way, if the area is spanning more than the visible window, the user can still get a sense of what is selected

**Comment by jaeilepp** [comment]
It is possible to deselect the area by re-clicking the corresponding line in the plot.
The area could be displayed as numbers left to the horizontal scrollbar, but it would overlap the old line. I'll see what I can do...

**Comment by mainakjas** [comment]
I see this in the terminal when I run the example:

```py
Traceback (most recent call last):
  File "/home/mainak/anaconda/lib/python2.7/site-packages/matplotlib/backends/backend_qt5.py", line 274, in mouseDoubleClickEvent
    button, dblclick=True)
  File "/home/mainak/anaconda/lib/python2.7/site-packages/matplotlib/backend_bases.py", line 1870, in button_press_event
    self.callbacks.process(s, mouseevent)
  File "/home/mainak/anaconda/lib/python2.7/site-packages/matplotlib/cbook.py", line 540, in process
    proxy(*args, **kwargs)
  File "/home/mainak/anaconda/lib/python2.7/site-packages/matplotlib/cbook.py", line 415, in __call__
    return mtd(*args, **kwargs)
  File "/home/mainak/Desktop/projects/mne-python/mne/viz/raw.py", line 198, in _mouse_click
    _pick_bad_channels(event, params)
  File "/home/mainak/Desktop/projects/mne-python/mne/viz/raw.py", line 129, in _pick_bad_channels
    this_chan = vars(l)['ch_name']
KeyError: 'ch_name'

```

**Comment by mainakjas** [comment]
On Tue, May 5, 2015 at 3:11 PM, jaeilepp <notifications@github.com> wrote:

> It is possible to deselect the area by re-clicking the corresponding line
> in the plot.
>
Ah, cool!

> The area could be displayed as numbers left to the horizontal scrollbar,
> but it would overlap the old line. I'll see what I can do...
>
Use two different color and put them one below the other? Maybe orange for
only the vertical line and green when selecting the area?

**Comment by mainakjas** [comment]
There is one other usability issue for me. What do the shift + left and right mouse clicks mean? Currently, they can control both the "start" and "end" of the area. There is a "flipping" effect if the user clicks beyond the "end" with the shift + left click. This to me is confusing. Actually, do we even need two different controls for the "start" and "end"?

**Comment by mainakjas** [comment]
We should also add tests

**Comment by jaeilepp** [comment]
Well, I thought it would be more usable if right click marks the end and left click marks the start. That way we would have more control over the lines. Otherwise it would be possible to add more lines and the logic would be broken.
The issue about setting the 'start' point after the 'end' point is negotiable. In my opinion it doesn't really matter, and it is up to the user to select the area any way he/she wants. It is pretty straightforward to determine the painted area either way.

**Comment by jaeilepp** [comment]
It might be good to make the lines in different colors so it would be easier to differentiate them.

**Comment by mainakjas** [comment]
I think I like the fact that the "start" and "end" point are in the same color. Or do you mean the vertical line? That should probably be in a different color.

File: mne/viz/raw.py [hunk 1]
```diff
@@ -163,7 +164,38 @@ def _mouse_click(event, params):
...
+                    return
+                params['ax_leftline'].set_data(x, y)
+                xdata = params['ax_rightline'].get_xdata()
+                params['fillarea'] = params['ax'].fill_betweenx(y, x, xdata,
+                                                                facecolor='y',
+                                                                alpha=0.2)
+            elif event.button == 3:
+                xdata = params['ax_rightline'].get_xdata()
+                if np.abs(xdata[0] - x[0]) < 0.2:
+                    x = params['raw'].index_as_time(params['raw'].last_samp)[0]
```
Comment (mainakjas) [inline mne/viz/raw.py]: this variable name is very uninformative

File: mne/viz/raw.py [hunk 2]
```diff
@@ -163,7 +164,38 @@ def _mouse_click(event, params):
         _plot_raw_time(event.xdata - params['duration'] / 2, params)
 
     elif event.inaxes == params['ax']:
-        _pick_bad_channels(event, params)
+        # vertical lines added
+        if event.key == 'shift':
+            if (params['fillarea'] in params['ax'].collections):
+                params['ax'].collections.remove(params['fillarea'])
+            x = np.array([event.xdata] * 2)
+            y = np.array(params['ax'].get_ylim())
```
Comment (mainakjas) [inline mne/viz/raw.py]: same here

**Comment by jaeilepp** [comment]
I don't know what you mean with points, but I meant the start and end lines that delimit the painted area. They don't really show up in the figure that well...

**Comment by mainakjas** [comment]
okay, I am happy with the delimiting lines. Let me know when you have made the fixes and I can review again.

**Comment by dengemann** [comment]
@mainakjas ping me when you're don, then I will take a look. Hi + welcome @jaeilepp :)

**Comment by mainakjas** [comment]
@jaeilepp don't hesitate to comment on other pull requests. The whole point of community bonding is to interact with the community as much as possible :)

**Comment by jaeilepp** [comment]
Now it paints the area also at the scrollbar. The time points are shown as text next to the scrollbar.

**Comment by mainakjas** [comment]
... almost there. But now the problem is that when I click near one of the delimiting lines to unselect the area, I have two vertical lines left. I would like to remove both the delimiting lines when I unselect the area.

**Comment by mainakjas** [comment]
I see that you haven't updated any documentation. How is the user supposed to know that they should do `shift + left-click` or `shift + right-click` ?

**Comment by jaeilepp** [comment]
Which documentation do you mean exactly? Is it enough to add a line to what's new?

**Comment by mainakjas** [comment]
No, I mean the documentation of the function. The docstring. I think in the `Notes` section but you should check

**Comment by jaeilepp** [comment]
I didn't quite understand the first comment. You delete the 'first' line with left click and the 'second' line with a right click...
I don't think It's a good idea to delete both lines with one click, but maybe there could be a shortcut key for that too.

**Comment by mainakjas** [comment]
okay, I see. So, we need two clicks to get rid of the area? Can't we have `shift + left-click` to create the area and `shift + right-click` to unselect it? Too many controls may confuse the user ...

**Comment by mainakjas** [comment]
Basically, to me the most intuitive way is to keep the mouse button pressed as I make the area and then finish it when I release the mouse button. But I am not sure how hard it is to implement.

**Comment by jaeilepp** [comment]
Yes that would work, but it would be difficult to paint larger areas than the axes width (100s of seconds on a 10s window).

**Comment by jaeilepp** [comment]
We still haven't decided what the user can do with the painted area. At the moment it's just a visual trick. And about the hangout session. Wasn't it supposed to be going on now?

**Comment by mainakjas** [comment]
Check your inbox. Do you have a gmail email address?

**Comment by jaeilepp** [comment]
Yes, this is a gmail address, but it might be that hangout is disabled on
university email. Is skype ok?

File: mne/viz/raw.py [hunk 3]
```diff
@@ -403,6 +453,10 @@ def plot_raw(raw, events=None, duration=10.0, start=0.0, n_channels=None,
     To mark or un-mark a channel as bad, click on the rather flat segments
     of a channel's time series. The changes will be reflected immediately
     in the raw object's ``raw.info['bads']`` entry.
+    Shift + mouse click can be used to paint an area in the plot. Shift +
+    left click marks the beginning of an area and shift + right click marks the
+    end of the area. The area and the markers can be deleted by clicking the
+    start and end points again.
```
Comment (mainakjas) [inline mne/viz/raw.py]: you added this to a private function but the user won't see this.

**Comment by agramfort** [comment]
does not seem to work on my mac book. Did not test with external mouse but it should work without.

**Comment by jaeilepp** [comment]
That's strange. I have no possibility to try it with mac. Have you tried changing the matplotlib back end to TkAgg? It seems to work with Qt back end too...

**Comment by agramfort** [comment]
even with an external mouse it does not work on my mac. Maybe I am doing something wrong...

@mainakjas do me a demo when you're around

**Comment by Eric89GXL** [comment]
@agramfort does the current `master` vertical line and bad-channel marking functionality work for you?

**Comment by agramfort** [comment]
Yes

**Comment by mainakjas** [comment]
any other mac users here? @teonlamont do you want to give it a try?

**Comment by Eric89GXL** [comment]
@jaeilepp I don't think it's a backend thing because the current mouse-based interactivity works on `master` for @agramfort, but the mouse-based interactivity in this PR doesn't. Is there something you're doing differently in terms of blocking/non-blocking or event handling?

File: mne/viz/raw.py [hunk 4]
```diff
@@ -163,7 +164,56 @@ def _mouse_click(event, params):
...
+                params['ax_hscroll_rightline'].set_data(x, [0, 1])
+                xdata = params['ax_leftline'].get_xdata()
+                params['fillarea'] = params['ax'].fill_betweenx(y, xdata, x,
+                                                                alpha=0.2,
+                                                                facecolor='y')
+                params['hscroll_fillarea'] = scrl.fill_betweenx([0, 1], x,
+                                                                xdata,
+                                                                alpha=0.2,
+                                                                facecolor='y')
+            event.canvas.draw()
```
Comment (Eric89GXL) [inline mne/viz/raw.py]: This should probably all be put into a private function like `_pick_bad_channels` is to keep this logical separation clean

**Comment by Eric89GXL** [comment]
I got out my mac and investigated. It looks like with the default backend (I assume Qt) in the Python interpreter the button-presses don't go to matplotlib at all, even if `raw.plot(block=True)` is used. Argh... not sure what a good workaround would be.

**Comment by Eric89GXL** [comment]
Related reading: https://github.com/matplotlib/matplotlib/issues/665/

**Comment by jaeilepp** [comment]
I suspected it might be related to that. I also noticed that writing tests with a modifier key ('shift') is a bit tricky.

**Comment by agramfort** [comment]
left click works fine for me to get the green bar. It's just shift-click
and the area that fails

**Comment by mainakjas** [comment]
So, it's just a problem because we use the modifier key `shift`? Why not replace it with click + drag + unclick sequence? It's more intuitive to me anyway.

**Comment by agramfort** [comment]
i like this if it works :)

**Comment by mainakjas** [comment]
And to paint areas larger than the axis width, the user just goes to the end of the axis and it should automatically start scrolling. You see what I mean?

**Comment by agramfort** [comment]
Can I close this?

**Comment by jaeilepp** [comment]
> Can I close this?

Fine by me.

**Pull request closed without merging**